File: ConsumerVPN-Android/consumerVPN/src/main/java/com/wlvpn/slider/whitelabelvpn/jobs/ServersRefreshJob.java
package com.wlvpn.slider.whitelabelvpn.jobs;

import android.support.annotation.NonNull;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobRequest;
import com.wlvpn.slider.whitelabelvpn.BuildConfig;
import com.wlvpn.slider.whitelabelvpn.ConsumerVpnApplication;
import com.wlvpn.slider.whitelabelvpn.managers.AccountManager;

import kotlin.Unit;
import timber.log.Timber;

/**
 * Job implementation for updating the server list.
 */
public class ServersRefreshJob extends Job {

    /**
     * Job ID
     */
    static final String JOB_ID = "ServersRefreshJob";

    private final AccountManager accountManager;

    ServersRefreshJob(AccountManager accountManager) {
        this.accountManager = accountManager;
    }

    /**
     * Schedules the job
     */
    public static void schedule() {
        //For debug builds use reduced timers and more accepting connection type
        long start = BuildConfig.DEBUG ?
                ConsumerVpnJobCreator.FIFTEEN_MINUTE_IN_MILLISECONDS
                : ConsumerVpnJobCreator.THIRTY_MINUTES_MILLISECOND;

        JobRequest.NetworkType networkType = BuildConfig.DEBUG ?
                JobRequest.NetworkType.CONNECTED
                : JobRequest.NetworkType.UNMETERED;

        new JobRequest.Builder(JOB_ID)
                .setRequiredNetworkType(networkType)
                .setRequirementsEnforced(true)
                .setPeriodic(start)
                .setUpdateCurrent(true)
                .build()
                .schedule();
    }

    @NonNull
    @Override
    protected Result onRunJob(@NonNull Params params) {

        if (accountManager.isUserLoggedIn()) {

            // Blocking current thread to allow the refresh token refresh finish
            // before sending Result.Success to Job
            ConsumerVpnApplication.getVpnSdk()
                    .updateServerList()
                    .blockCurrentThread()
                    .subscribe(
                            vpnServersResponse -> Unit.INSTANCE,
                            throwable -> {
                                Timber.e(throwable, "Servers refresh failed during job");
                                return Unit.INSTANCE;
                            });
        }

        return Result.SUCCESS;
    }
}

File: ConsumerVPN/app/src/main/java/com/wlvpn/consumervpn/data/job/ServersRefreshJob.kt
package com.wlvpn.consumervpn.data.job

import com.evernote.android.job.Job
import com.evernote.android.job.JobRequest
import com.wlvpn.consumervpn.BuildConfig
import com.wlvpn.consumervpn.domain.service.servers.ServersService
import timber.log.Timber

class ServersRefreshJob internal constructor(
    private val serversService: ServersService
) : Job() {

    companion object {

        internal const val JOB_ID = "consumer:SeversRefreshJob"

        fun schedule() {
            //For debug builds use reduced timers and more accepting connection listType
            val start =
                if (BuildConfig.DEBUG) FIFTEEN_MINUTES_IN_MILLISECONDS
                else TWO_HOURS_IN_MILLISECONDS

            JobRequest.Builder(JOB_ID)
                .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                .setRequirementsEnforced(true)
                .setPeriodic(start)
                .setUpdateCurrent(true)
                .build()
                .schedule()
        }
    }

    override fun onRunJob(params: Job.Params): Job.Result {

        serversService.refreshServers()
            .blockingGet()
            ?.let { throwable ->
                Timber.e(throwable, "Error while refreshing current servers in servers refresh job")
            }

        return Job.Result.SUCCESS
    }
}
